diff --git a/waakama/examples/shared/connection.h b/examples/shared/connection.h
index c676bac..f8bd293 100644
--- a/wakaama/examples/shared/connection.h
+++ b/wakaama/examples/shared/connection.h
@@ -18,6 +18,11 @@
 #ifndef CONNECTION_H_
 #define CONNECTION_H_
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
 #include <stdio.h>
 #include <unistd.h>
 #include <netinet/in.h>
@@ -52,4 +57,8 @@ void connection_free(connection_t * connList);
 
 int connection_send(connection_t *connP, uint8_t * buffer, size_t length);
 
+#ifdef __cplusplus
+}
+#endif
+
 #endif
